File: server/src/core/topic_hasher.py
"""
키워드 → FCM Topic 해시 변환

iOS(Swift)와 Python 양쪽에서 동일한 해시를 생성해야 합니다.
정규화 순서: strip() → lower() → SHA256 → hexdigest[:16]
"""

from __future__ import annotations

import hashlib


def topic_name(keyword: str, noti_type: str = "bid") -> str:
    """키워드를 FCM Topic 이름으로 변환

    FCM Topic 이름 규칙: [a-zA-Z0-9-_.~%]{1,900}
    한글 키워드를 SHA256 해시로 변환합니다.

    Args:
        keyword: 원본 키워드 (예: "CCTV", "소프트웨어")
        noti_type: "bid" (입찰공고) 또는 "pre" (사전규격)

    Returns:
        FCM Topic 이름 (예: "bid_682608e4a04d8e01")
    """
    prefix = "bid_" if noti_type == "bid" else "pre_"
    normalized = keyword.strip().lower()
    hex_hash = hashlib.sha256(normalized.encode("utf-8")).hexdigest()[:16]
    return f"{prefix}{hex_hash}"


def get_all_topic_names(keyword: str) -> dict[str, str]:
    """키워드의 입찰공고/사전규격 토픽 이름을 모두 반환

    Args:
        keyword: 원본 키워드

    Returns:
        {"bid_topic": "bid_...", "pre_topic": "pre_..."}
    """
    return {
        "bid_topic": topic_name(keyword, "bid"),
        "pre_topic": topic_name(keyword, "pre"),
    }

File: BidAlert/Models/TopicHasher.swift
import CryptoKit
import Foundation

/// 키워드 → FCM Topic 해시 변환
///
/// ⚠️ 서버(Python)의 `topic_hasher.py`와 반드시 동일한 결과를 생성해야 합니다.
///
/// 정규화 순서: trimmingCharacters → lowercased → SHA256 → hex prefix(16)
///
/// 검증 기준값:
/// - "cctv"     → "b29dbba57df61de7"
/// - "소프트웨어" → "465f222a27475e7f"
/// - "ai"       → "32e83e92d45d71f6"
/// - "측량"     → "5f66b02e337d9504"
enum TopicHasher {

    enum NotificationType: String {
        case bid = "bid"
        case prebid = "pre"
    }

    /// 키워드를 FCM Topic 이름으로 변환
    /// - Parameters:
    ///   - keyword: 원본 키워드 (예: "CCTV", "소프트웨어")
    ///   - type: 알림 유형 (.bid 또는 .prebid)
    /// - Returns: FCM Topic 이름 (예: "bid_b29dbba57df61de7")
    static func topicName(for keyword: String, type: NotificationType) -> String {
        let prefix = type == .bid ? "bid_" : "pre_"
        let normalized = keyword.trimmingCharacters(in: .whitespaces).lowercased()
        let hash = SHA256.hash(data: Data(normalized.utf8))
        let hex = hash.compactMap { String(format: "%02x", $0) }.joined()
        let truncated = String(hex.prefix(16))
        return "\(prefix)\(truncated)"
    }

    /// 키워드의 입찰공고 + 사전규격 토픽 이름 동시 반환
    static func allTopics(for keyword: String) -> (bid: String, pre: String) {
        return (
            bid: topicName(for: keyword, type: .bid),
            pre: topicName(for: keyword, type: .prebid)
        )
    }
}
